cvat-core/src/math/index.js

```diff
@@ -0,0 +1,191 @@
+const { Matrix, solve: leastSquaresMethod } = require("ml-matrix");
+
+function findVanishingPoint(lineSegments, infinityDistance) {
+  // find the vanishing point and update line segments to intersect
+  // exactly at the vanishing point. If vanishing point cannot be found
+  // (parallel lines), segments are updated to be exactly parallel
+  const segmentCenters = lineSegments.map((s) => getMidPoint(s[0], s[1]));
+  const lines = lineSegments.map((s) => getLineByTwoPoints(s[0], s[1]));
+  let vanishingPoint = approximateLinesIntersection(lines);
+  if (vanishingPoint && !isFinite(vanishingPoint, infinityDistance)) {
+    vanishingPoint = null;
+  }
+  let adjustedLines;
+  if (vanishingPoint) {
+    adjustedLines = segmentCenters.map((center) =>
+      getLineByTwoPoints(center, vanishingPoint)
+    );
+  } else {
+    const angleFactor = approximateAngleFactor(lines);
+    adjustedLines = segmentCenters.map((point) =>
+      buildLineThrough(point, angleFactor)
+    );
+  }
+  const adjustedLineSegments = lineSegments.map((s, i) => [
+    projectOntoLine(s[0], adjustedLines[i]),
+    projectOntoLine(s[1], adjustedLines[i])
+  ]);
+  return [adjustedLineSegments, vanishingPoint];
+}
+
+function updateSegment(previous, next, vanishingPoint) {
+    const [prevA, prevB] = previous;
+    const [nextA, nextB] = next;
+
+    if (vanishingPoint) {
+        if (!arePointsEqual(prevA, nextA)) {
+            const line = getLineByTwoPoints(vanishingPoint, nextA);
+            const newB = projectOntoLine(prevB, line);
+            return [nextA, newB];
+        } else if (!arePointsEqual(prevB, nextB)) {
+            const line = getLineByTwoPoints(vanishingPoint, nextB);
+            const newA = projectOntoLine(prevA, line);
+            return [newA, nextB];
+        }
+        return next;
+    }
+
+    let line = getLineByTwoPoints(prevA, prevB);
+    if (!arePointsEqual(prevA, nextA)) {
+        line = buildLineThrough(nextA, line);
+        const newB = projectOntoLine(prevB, line);
+        return [nextA, newB];
+    } else if (!arePointsEqual(prevB, nextB)) {
+        line = buildLineThrough(nextB, line);
+        const newA = projectOntoLine(prevA, line);
+        return [nextA, newA];
+    }
+    return next;
+}
+
+function arePointsEqual(first, second) {
+    return Math.abs(first.x - second.x) < 0.001 && Math.abs(first.y - second.y) < 0.001;
+}
+
+function pullSegment(point, previousEdges, nextEdges) {
+  const [previousStart, previousEnd] = previousEdges;
+  const [nextStart, nextEnd] = nextEdges;
+  const [ratioX, lengthX] = _calcRatio(previousStart.x, point.x, previousEnd.x);
```

Comment:
if you want to calculate ratio of a point over a line, the easiest is to use dot product:
(P - A)*(B - A) = ratio * norm(B - A)**2

Generally, you will save yourself some trouble if you have a mini-module to do math on 2D points, maybe CVAT already has it?